### ISSUE TITLE
appendManifest API is not thread safe

### ISSUE THREAD

**Issue by puchengy** [apache/iceberg#8420]
### Feature Request / Improvement

By looking at the implementation of `appendManifest` method, it is using data structure that is not thread safe, for example ```
private final List<ManifestFile> rewrittenAppendManifests = Lists.newArrayList();
```
Will be great to make it thread safe such that we can add manifests in parallel.

### Query engine

None

**Comment by steinsgateted** [comment]
Hello @puchengy @Fokko, 
I would like to take this issue.
I would like some clarifications.
Does it mean to make the data structure thread-safe?
Like this:
`private final List<ManifestFile> rewrittenAppendManifests = Lists.newArrayList();`
change to:
`private final List<ManifestFile> rewrittenAppendManifests = new CopyOnWriteArrayList<>();`

**Comment by puchengy** [comment]
@steinsgateted Yes, correct. TBH, once we make it thread safe, I am not sure how to ensure people not introducing new thread unsafe data structure later on, do you know?
